### DISCUSSION TITLE
Emulate csvstack

### DISCUSSION THREAD

**Discussion by persep** [dathere/qsv#1053] [Q&A] [answered]
Hi,
is there anyway of using `qsv ` like the [csvstack](https://csvkit.readthedocs.io/en/latest/scripts/csvstack.html) comand? in particular I'm looking for a way of adding a grouping value to each row. Say we have these 2 csv files:

`letters.csv`
```
h1,h2
a, b
c, d
```

and `numbers.csv`

```
h1,h2
1, 2
3, 4
```

With `csvstack` I can join both files and add a new column that references to the csv file that was used

```
$ csvstack -g letter,number  -n type letters.csv numbers.csv
type,h1,h2
letter,a, b
letter,c, d
number,1, 2
number,3, 4
```

If it isn't possible I think it would be good to add it to the `qsv cat` command

**Comment by aborruso** [comment]
A small note: your example does not work.

I think you should edit `numbers.csv` header, to have `h1,h2` and not `h3,h4`.

**Reply by persep** [reply]
Yes it does work, I think it's because the number of columns is the same.
But as I said it's the grouping factor rather than the column names that I'm interested in

**Reply by persep** [reply]
I'm rewriting the example. It worked on my computer because I had an old version installed of csvkit

**Comment by aborruso** [comment]
> Yes it does work, I think it's because the number of columns is the same.

For me it does not work, I have 

```
+--------+----+----+----+----+
| type   | h1 | h2 | h3 | h4 |
+--------+----+----+----+----+
| letter | a  |  b | -  | -  |
| letter | c  |  d | -  | -  |
| number | -  | -  | 1  |  2 |
| number | -  | -  | 3  |  4 |
+--------+----+----+----+----+
```

I'm sure it can be possible to do in qsv, using cat and some other command that can map the name of the files, but it's better to have the @jqnatividad answer

**Comment by jqnatividad** [comment]
Hi @persep,

Thanks for the detailed request.  And thanks @aborruso for jumping in with some feedback.

Right now, nothing comes immediately to mind, but I'm sure a `luau` recipe can be whipped up to do something like `csvstack`.

FYI, when I integrated `luau`, I was hoping that a library of reusable, community-contributed recipes will be compiled in the Wiki section.  However, that has not happened, and I intend to address that later this year by seeding a searchable luau recipe library and making it easy to discover, contribute & reuse recipes from the library.

But back to the question at hand, I too was a big csvkit user before I started using xsv (discovering it thru [csvkit's About](https://csvkit.readthedocs.io/en/latest/index.html#about) page), and then eventually created the qsv fork.

So for now, I'm inclined to create a csvstack analogue - either with a luau recipe or a new command., so I'll add this to the enhancement backlog - https://github.com/jqnatividad/qsv/issues/1057

**Reply by aborruso** [reply]
> FYI, when I integrated luau, I was hoping that a library of reusable, community-contributed recipes will be compiled in the Wiki section.

You are right, I would help (as I do with other tools like Miller or VisiData), but I have no idea how lua works.

I just wanted to tell you that I'm sorry, because you do a great job and I wish I could be more able to help you

**Reply by persep** [reply]
The curious thing is that the only tool so far that that I've found that does that grouping is the `csvstack`
While looking for a solution I found this [suggestion](https://github.com/BurntSushi/xsv/issues/16#) in the xsv github. The last [comment](https://github.com/BurntSushi/xsv/issues/16#issuecomment-1385562904) suggests using a cross join with another csv file that would contain the grouping element and then all you would have to do is merge all the csv files but as it says in the comment you still need to make a new csv file file for it to work.

**Reply by aborruso** [reply]
> The curious thing is that the only tool so far that that I've found that does that grouping is the csvstack

@persep  No, you can use in example Miller
https://stackoverflow.com/questions/76506396/generalize-a-awk-program-that-joins-csv-files-to-n-arguments/76511765#76511765

**Reply by persep** [reply]
But only if the name in files is the name of the group factors. IMHO  `csvstack`'s argument system is better

**Reply by aborruso** [reply]
> But only if the name in files is the name of the group factors. IMHO csvstack's argument system is better

Ok, here you are asking to emulate csvstack, then you are not asking a generic question, otherwise it would be a leading question.

In some way you want a row classification by file. So, you could start by inserting a column with the name of the file (as in AWK and Miller), and then using some "if conditions" assign the label you want.

**Comment by jqnatividad** [comment, marked as answer]
Hi @persep , @aborruso ,

I went ahead and implemented grouping by file name stem options in `cat rowskey`.

#1067 

Let me know what you think!

**Reply by persep** [reply]
It works great, good job

**Reply by aborruso** [reply]
I'm happy for it. To use the file name was for me the right way to do it.

Thank you @jqnatividad
